### DISCUSSION TITLE
Modbus connection issues

### DISCUSSION THREAD

**Discussion by kodebach** [kodebach/hacs-idm-heatpump#133] [General]
> Yes that is the one, I suffered many times this year. The guys ID is janiversen and seems to develop the pymodbus in the wrong direction. 
>
> I do not talk about HA integration or sematic versioning. I talk about modbusTCP functionality he broke several times this year and many users complaint. By the way he is good in removing issues by just closing them but not solving...
> 1. When using native modbus integration in HA, you need to define at least on entity. This was not the case last year and was the first breaking change. I use your integration for reading and juse use the native one for writing a few registers when needed. This can't be understood by that guy...
> 2. Now after restart, you have to wait a little time. Before pymodbus synchronized that themselves, now gone...
> 3. restart will disappear in Nov. I have no idea how to ensure a modbus write workes, because we know IDMs modbusTCP connection are not as reliable as we wanted it to be. A simple restart before write solved the issue till now....
>
> I am quite annoyed (as other users), this guy alsways states, it is against the spec. blabla...  I had a few discussions with him and others as well. It ends like, please do it yourself.... 
Of course he could say so, but why did he put effort in breaking a good working library by changing it? For me crazy. He ignores, modbusTCP devices are not always working against spec, The previous pymodbus implementation gave you workarounds like the restart, which will disappear in Nov. this year.... Oh my good....

_Originally posted by @AndyNew2 in https://github.com/kodebach/hacs-idm-heatpump/issues/126#issuecomment-2333470111_

@AndyNew2 I moved this here, so we don't pollute the issue with off topic discussion.

**Comment by kodebach** [comment]
> When using native modbus integration in HA, you need to define at least on entity.

This has nothing to do with `pymodbus`. That's a decision from whoever maintains the modbus integration in HA (might be the same person, no idea the [`manifest.json`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/modbus/manifest.json) doesn't list anyone). It also sounds more like a decision from HA itself, since using an integration in HA without defining an entity or service doesn't make all that much sense. 


> I have no idea how to ensure a modbus write workes, because we know IDMs modbusTCP connection are not as reliable as we wanted it to be.

That is clearly an issue for IDM. Also, ensuring writes work is a very simply issue. You can simply retry repeatedly until it works or you give up. Every Modbus request (whether read or write) should result in a response (from the heat pump), so it is very clear whether the request worked or not.

> this guy alsways states, it is against the spec.

Conforming to the spec is very good reasoning. Of course they could be nicer about it, but I don't blame them for wanting to keep their library minimal.

> He ignores, modbusTCP devices are not always working against spec

Again there is a spec, if a device doesn't conform to the spec, it's an issue with the device. 

> The previous pymodbus implementation gave you workarounds like the restart

Not really, no. It was the workaround you used, but not something that was ever recommended by anyone AFAICT. AFAIK the `modbus.restart` service is also semi-broken, since it cannot fully reload and restart the integration since that is not how Home Assistant works. It also can cause leaks and if called often memory usage of HA could increase. Also please keep the difference between [pymodbus](https://github.com/pymodbus-dev/pymodbus) and the [HA modbus integration](https://github.com/home-assistant/core/tree/master/homeassistant/components/modbus) in mind. `pymodbus` never had a "restart" feature.


---

To address your issue: Why are you even using `modbus.restart`? Just as a precaution in case the heat pump doesn't respond? The current `pymobus` library underneath both this integration and the HA modbus integration should already reconnect and retry on connection failures. AFAICT this retry feature works for the reads, if the heat pump disconnects.

**Comment by AndyNew2** [comment]
janiversen does both, HA integration and supports also pymodbus. I am happy he supports the library but do not understand, why he put effort in breaking changes and destroying workarounds. 
I do not agree with this "must be aline with spec." I know perfectly from my own business, even TCP/IP stacks never ever fully implements the spec. You can't expect everything works perfectly according spec, reality is always different. You need to try and fix until it actually works.

What matters in the end, if it works or not. That is the key. If it is inline with spec. but not working, no good for anything. Best is, it works inline with specs. As we know the iDM modbusTCP is not 100% realiable. I searched on the internet and found many devices with similar issues. All fine, as long you could simply reconnect and fix the isse with retries. And I will not exchange a 20K worth heatpump just because the modbusTCP implementation is not perfect, you don't either. ;-)

According earlier integration pymodbus did not reconnects by its own. That was the source for the restart command, because I would not find it anyhow... The restart command fixed it. If I can rely pymodbus does on its own now reconnects, I would be happy, but I have seen behaviour different. It stopped working after a while and was broken since than.

If there are memory leaks using restart, I propose to fix this leaks, rather than removing a working workaround...

**Reply by kodebach** [reply]
> I do not agree with this "must be aline with spec."

AFAICT he never said he's not doing something because it is against the spec. He said there is no "modbus reset" in the spec, because he was confused and didn't know you were talking about the `modbus.restart` HA service provided by the `modbus` integration.

> why he put effort in breaking changes and destroying workarounds.

He did not. Again, from reading issues and PRs in the HA repo, it seems the `modbus.restart` service was subtly broken and could cause issues. That's why it's being removed.

> As we know the iDM modbusTCP is not 100% realiable

If it matter to you, report that to IDM and tell them to fix their firmware. They are the company that sold you expensive equipment. janiversen is a volunteer working on an open source project. He owes you nothing.

> All fine, as long you could simply reconnect and fix the isse with retries.

And `pymodbus` and therefore the HA modbus integration should still do that automatically, even without using the `modbus.restart` service.

> According earlier integration pymodbus did not reconnects by its own.

The HA modbus integration does initiate the `pymobus` Modbus client with a config with 3 retries. See also:

https://github.com/home-assistant/core/blob/17994ff245cc32dd27a8455e49cf530feb5383b6/homeassistant/components/modbus/modbus.py#L291-L296

In an earlier version this number was configurable, but it was removed in https://github.com/home-assistant/core/pull/105024. I don't entirely understand the motivation behind removing it, but it shouldn't matter. AFAIK the IDM heat pump always responds und the first reconnect & retry.

Since `pymodbus` 3.6.7 (April 2024) the retry process also captures failed connections https://github.com/pymodbus-dev/pymodbus/pull/2113.

Please try your scripts without the `modbus.restart`. I'm pretty sure it should just work with the current HA version.

**Comment by AndyNew2** [comment]
Since pymodbus 3.6.7 (April 2024) the retry process also captures failed connections https://github.com/pymodbus-dev/pymodbus/pull/2113.
I think this is the key change for me. So hopefully I am fine without the restart now. However is pretty new and I had the scripts developed in 2023, so the behaviour changed since then.

In principle I do not like they decide everything for me and break compatiblity. For me there was no reason to remove that useful function and I still could imagine situations, to have control over the reconnect to increase the likeless of a subsequent read or write call. For my case (controlling the heat pump) this is OK but it may differ for others.

At least it gives me the option to upgrade again HA, since I thought I need to stay on the October release. (The removal was forecasted Nov. 2024).

The report to iDM, you know exactly, is useless. Even it is a 20K worth heatpump you are not allowed to contact anyone one the development team. Furthermore there may be restrictions by the manufactur as well, since they use software packages, which may have the issue and they suffer themselves for control. I have seen something like that may times on my carrier. That is why I really knock on the door and want you to understand, you can't force everything to work perfectly against spec. This is just unrealitic and will not lead to a successful route.

In the decribed issue, I do not know the sequence. A really robost implementation should retry read and writes until expected result is given or the retry counts are exceed. TCP can handle closed socket exception, in that case it should be fine. But there may exist "handing" sockets, still connected according TCP, but the other peer stops responding. A reconnect increases the chance to recover the situation. According my reading this was not done in issue 2113 and the old restart was "safer" in that kind....
That is what I want to express, you never can predict all circumstances. These are discovered by the users, and the more possibilities the library gives, the higher the chance for a workaround.

I tried to explain that to janiversen, but he seems resistant against it, and it seems to be a new view on programmers. But according my 20 year experience, this is wrong.

As you described they removed the option for the retries and fixed it to 3. This is nonsense as well.  Also I suffer currently from many integration removing the update cycle time and fix it to 60 sec. or even lower. I have no idea, why they do so, but I had to discover a lot to finally restore higher update rates. This do not affect your library, but others like power meter and solar system integrations....

**Reply by kodebach** [reply]
> The report to iDM, you know exactly, is useless. Even it is a 20K worth heatpump you are not allowed to contact anyone one the development team.

I've never had a bad experience with IDM support and they are one of the few companies that still list an email and even a phone number for support on their website.

**Comment by AndyNew2** [comment]
That was the case maybe 5 - 7 years ago. Try again now.
I had many tries, but never got a response back. The e-mail adress shown on idm is reverted to your local dealer. Only if he local dealer forwards the request you get an answer at all. You remember I had the confirmation about the docu is wrong about Zonemodul writes, since they do not mark the harming flash memory. I asked them later, if it is better to write to main module or zone module, ... silence.
I asked them to add flow meter and hot gas values to modbusTCP, because these are the really critical values, ... silence no response
I asked them to add better control for max and min power to control the heatpump better and more optimized, ... silence no response
I reported them many mistakes in their control logic, ... silence no response.

Please go ahead and try to request any of this messages. If you are successful, you will be my hero! But I am sure, there is no feedback anymore....
